File: app/src/main/java/br/com/jonathanzanella/myexpenses/receipt/EditReceiptActivity.java
package br.com.jonathanzanella.myexpenses.receipt;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.UiThread;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.CheckBox;
import android.widget.EditText;

import org.apache.commons.lang3.StringUtils;
import org.joda.time.DateTime;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.account.Account;
import br.com.jonathanzanella.myexpenses.account.AccountRepository;
import br.com.jonathanzanella.myexpenses.account.ListAccountActivity;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyTextWatch;
import br.com.jonathanzanella.myexpenses.log.Log;
import br.com.jonathanzanella.myexpenses.source.ListSourceActivity;
import br.com.jonathanzanella.myexpenses.source.Source;
import br.com.jonathanzanella.myexpenses.source.SourceRepository;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;
import br.com.jonathanzanella.myexpenses.views.BaseActivity;
import butterknife.Bind;
import butterknife.OnClick;

public class EditReceiptActivity extends BaseActivity implements ReceiptContract.EditView {
	public static final String KEY_RECEIPT_UUID = "KeyReceiptUuid";
	private static final int REQUEST_SELECT_SOURCE = 1001;
	private static final int REQUEST_SELECT_ACCOUNT = 1002;

	@Bind(R.id.act_edit_receipt_name)
	EditText editName;
	@Bind(R.id.act_edit_receipt_date)
	EditText editDate;
	@Bind(R.id.act_edit_receipt_income)
	EditText editIncome;
	@Bind(R.id.act_edit_receipt_source)
	EditText editSource;
	@Bind(R.id.act_edit_receipt_account)
	EditText editAccount;
	@Bind(R.id.act_edit_receipt_repetition)
	EditText editRepetition;
	@Bind(R.id.act_edit_receipt_installment)
	EditText editInstallment;
	@Bind(R.id.act_edit_receipt_show_in_resume)
	CheckBox checkShowInResume;

	private final ReceiptPresenter presenter;

	public EditReceiptActivity() {
		presenter = new ReceiptPresenter(new ReceiptRepository(new RepositoryImpl<Receipt>(this)),
				new SourceRepository(new RepositoryImpl<Source>(this)),
				new AccountRepository(new RepositoryImpl<Account>(this)));
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_edit_receipt);
	}

	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);
		editIncome.addTextChangedListener(new CurrencyTextWatch(editIncome));
		presenter.attachView(this);
		presenter.viewUpdated(false);
	}

	@Override
	protected void storeBundle(Bundle extras) {
		super.storeBundle(extras);

		if(extras != null)
			presenter.storeBundle(extras);
	}

	@Override
	protected void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		presenter.onSaveInstanceState(outState);
	}

	@Override
	protected void onStart() {
		super.onStart();
		presenter.attachView(this);
	}

	@Override
	protected void onResume() {
		super.onResume();
		presenter.updateView();
	}

	@Override
	protected void onStop() {
		super.onStop();
		presenter.detachView();
	}

	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);
		presenter.attachView(this);
		switch (requestCode) {
			case REQUEST_SELECT_SOURCE: {
				if(resultCode == RESULT_OK) {
					String sourceUUid = data.getStringExtra(ListSourceActivity.KEY_SOURCE_SELECTED_UUID);
					if(sourceUUid != null)
						presenter.onSourceSelected(sourceUUid);
				}
				break;
			}
			case REQUEST_SELECT_ACCOUNT: {
				if(resultCode == RESULT_OK) {
					String accountUuid = data.getStringExtra(ListAccountActivity.Companion.getKEY_ACCOUNT_SELECTED_UUID());
					if(accountUuid != null)
						presenter.onAccountSelected(accountUuid);
				}
				break;
			}
		}
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.save, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.action_save:
				save();
				break;
		}
		return super.onOptionsItemSelected(item);
	}

	@OnClick(R.id.act_edit_receipt_date)
	void onDate() {
		presenter.onDate(this);
	}

	@Override
	public void onDateChanged(DateTime date) {
		editDate.setText(Receipt.SIMPLE_DATE_FORMAT.format(date.toDate()));
	}

	@Override
	public void onSourceSelected(Source source) {
		editSource.setText(source.getName());
	}

	@Override
	public void onAccountSelected(Account account) {
		editAccount.setText(account.getName());
	}

	@Override
	public int getInstallment() {
		return Integer.parseInt(editInstallment.getText().toString());
	}

	@Override
	public int getRepetition() {
		return Integer.parseInt(editRepetition.getText().toString());
	}

	@OnClick(R.id.act_edit_receipt_source)
	void onSource() {
		startActivityForResult(new Intent(this, ListSourceActivity.class), REQUEST_SELECT_SOURCE);
	}

	@OnClick(R.id.act_edit_receipt_account)
	void onAccount() {
		if(!presenter.hasReceipt())
			startActivityForResult(new Intent(this, ListAccountActivity.class), REQUEST_SELECT_ACCOUNT);
	}

	private void save() {
		presenter.save();
	}

	@Override
	public Receipt fillReceipt(Receipt receipt) {
		receipt.setName(editName.getText().toString());
		String income = editIncome.getText().toString().replaceAll("[^\\d]", "");
		if(!StringUtils.isEmpty(income))
			receipt.setIncome(Integer.parseInt(income));
		receipt.setShowInResume(checkShowInResume.isChecked());
		receipt.setInstallments(getInstallment());
		receipt.setRepetition(getRepetition());
		return receipt;
	}

	@Override
	public void finishView() {
		Intent i = new Intent();
		i.putExtra(KEY_RECEIPT_UUID, presenter.getUuid());
		setResult(RESULT_OK, i);
		finish();
	}

	@Override
	public void showError(ValidationError error) {
		switch (error) {
			case NAME:
				editName.setError(getString(error.getMessage()));
				break;
			case AMOUNT:
				editIncome.setError(getString(error.getMessage()));
				break;
			case SOURCE:
				editSource.setError(getString(error.getMessage()));
				break;
			case ACCOUNT:
				editAccount.setError(getString(error.getMessage()));
				break;
			default:
				Log.error(this.getClass().getName(), "Validation unrecognized, field:" + error);
		}
	}

	@Override
	@UiThread
	public void showReceipt(final Receipt receipt) {
		editName.setText(receipt.getName());
		editIncome.setText(CurrencyHelper.format(receipt.getIncome()));
		if(receipt.isCredited())
			//noinspection deprecation
			editIncome.setTextColor(getResources().getColor(R.color.value_unpaid));

		new AsyncTask<Void, Void, Source>() {

			@Override
			protected Source doInBackground(Void... voids) {
				return receipt.getSource();
			}

			@Override
			protected void onPostExecute(Source source) {
				super.onPostExecute(source);
				editSource.setText(source.getName());
			}
		}.execute();

		checkShowInResume.setChecked(receipt.isShowInResume());
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/receipt/EditReceiptActivity.kt
package br.com.jonathanzanella.myexpenses.receipt

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.annotation.UiThread
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.AppCompatEditText
import android.text.InputType
import android.view.Menu
import android.view.MenuItem
import android.widget.CheckBox
import br.com.jonathanzanella.myexpenses.App
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.account.ListAccountActivity
import br.com.jonathanzanella.myexpenses.helpers.CurrencyTextWatch
import br.com.jonathanzanella.myexpenses.helpers.toCurrencyFormatted
import br.com.jonathanzanella.myexpenses.source.ListSourceActivity
import br.com.jonathanzanella.myexpenses.source.Source
import br.com.jonathanzanella.myexpenses.transaction.Transaction
import br.com.jonathanzanella.myexpenses.validations.ValidationError
import br.com.jonathanzanella.myexpenses.views.anko.*
import org.apache.commons.lang3.StringUtils
import org.jetbrains.anko.*
import org.joda.time.DateTime
import timber.log.Timber
import javax.inject.Inject

class EditReceiptActivity : AppCompatActivity(), ReceiptContract.EditView {
    @Inject
    lateinit var presenter: ReceiptPresenter
    override val context = this
    private val ui = EditReceiptActivityUi()

    override val installment: Int
        get() = Integer.parseInt(ui.editInstallment.text.toString())
    override val repetition: Int
        get() = Integer.parseInt(ui.editRepetition.text.toString())

    init {
        App.getAppComponent().inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ui.setContentView(this)

        storeBundle(savedInstanceState)
        storeBundle(intent.extras)

        ui.toolbar.setup(this)
        ui.editIncome.addTextChangedListener(CurrencyTextWatch(ui.editIncome))

        presenter.attachView(this)
        presenter.viewUpdated(false)
    }

    fun storeBundle(extras: Bundle?) {
        if (extras != null)
            presenter.storeBundle(extras)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        presenter.onSaveInstanceState(outState)
    }

    override fun onStart() {
        super.onStart()
        presenter.attachView(this)
    }

    override fun onResume() {
        super.onResume()
        presenter.updateView()
    }

    override fun onStop() {
        super.onStop()
        presenter.detachView()
    }

    override fun setTitle(string: String) {
        ui.toolbar.title = string
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        presenter.attachView(this)
        when (requestCode) {
            REQUEST_SELECT_SOURCE -> {
                if (resultCode == Activity.RESULT_OK) {
                    val sourceUUid = data!!.getStringExtra(ListSourceActivity.KEY_SOURCE_SELECTED_UUID)
                    if (sourceUUid != null)
                        presenter.onSourceSelected(sourceUUid)
                }
            }
            REQUEST_SELECT_ACCOUNT -> {
                if (resultCode == Activity.RESULT_OK) {
                    val accountUuid = data!!.getStringExtra(ListAccountActivity.KEY_ACCOUNT_SELECTED_UUID)
                    if (accountUuid != null)
                        presenter.onAccountSelected(accountUuid)
                }
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.save, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_save -> save()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDateChanged(balanceDate: DateTime) {
        ui.editDate.setText(Transaction.SIMPLE_DATE_FORMAT.format(balanceDate.toDate()))
    }

    override fun onSourceSelected(source: Source) {
        ui.editSource.setText(source.name)
    }

    override fun onAccountSelected(account: Account) {
        ui.editAccount.setText(account.name)
    }

    internal fun onDate() {
        presenter.onDate(this)
    }

    internal fun onSource() {
        startActivityForResult(Intent(this, ListSourceActivity::class.java), REQUEST_SELECT_SOURCE)
    }

    internal fun onAccount() {
        if (!presenter.hasReceipt())
            startActivityForResult(Intent(this, ListAccountActivity::class.java), REQUEST_SELECT_ACCOUNT)
    }

    private fun save() {
        presenter.save()
    }

    override fun fillReceipt(receipt: Receipt): Receipt {
        receipt.name = ui.editName.text.toString()
        val income = ui.editIncome.text.toString().replace("[^\\d]".toRegex(), "")
        if (!StringUtils.isEmpty(income))
            receipt.income = Integer.parseInt(income)
        receipt.isShowInResume = ui.checkShowInResume.isChecked
        receipt.installments = installment
        receipt.repetition = repetition
        return receipt
    }

    override fun showConfirmDialog(receipt: Receipt) {
        AlertDialog.Builder(ctx)
                .setMessage("${getString(R.string.message_confirm_receipt)} ${receipt.name} - ${receipt.incomeFormatted}?")
                .setPositiveButton(R.string.yes) { _, _ ->
                    doAsync {
                        receipt.credit()

                        uiThread { finishView() }
                    }
                }
                .setNegativeButton(R.string.no) { dialogInterface, _ ->
                    dialogInterface.dismiss()
                    finishView()
                }
                .show()
    }

    override fun finishView() {
        val i = Intent()
        i.putExtra(KEY_RECEIPT_UUID, presenter.uuid)
        setResult(Activity.RESULT_OK, i)
        finish()
    }

    override fun showError(error: ValidationError) {
        when (error) {
            ValidationError.NAME -> ui.editName.error = getString(error.message)
            ValidationError.AMOUNT -> ui.editIncome.error = getString(error.message)
            ValidationError.SOURCE -> ui.editSource.error = getString(error.message)
            ValidationError.ACCOUNT -> ui.editAccount.error = getString(error.message)
            else -> Timber.e("Validation unrecognized, field:" + error)
        }
    }

    @UiThread
    override fun showReceipt(receipt: Receipt) {
        ui.editName.setText(receipt.name)
        ui.editIncome.setText(receipt.income.toCurrencyFormatted())
        if (receipt.credited)
            ui.editIncome.setTextColor(ResourcesCompat.getColor(resources, R.color.value_unpaid, null))

        doAsync {
            val source = receipt.source

            uiThread { ui.editSource.setText(source?.name) }
        }

        ui.checkShowInResume.isChecked = receipt.isShowInResume
    }

    companion object {
        val KEY_RECEIPT_UUID = "KeyReceiptUuid"
        private val REQUEST_SELECT_SOURCE = 1001
        private val REQUEST_SELECT_ACCOUNT = 1002
    }
}

private class EditReceiptActivityUi : AnkoComponent<EditReceiptActivity> {
    lateinit var toolbar : TemplateToolbar
    lateinit var editName : AppCompatEditText
    lateinit var editDate: AppCompatEditText
    lateinit var editIncome: AppCompatEditText
    lateinit var editSource: AppCompatEditText
    lateinit var editAccount: AppCompatEditText
    lateinit var checkShowInResume: CheckBox
    lateinit var editRepetition: AppCompatEditText
    lateinit var editInstallment: AppCompatEditText

    override fun createView(ui: AnkoContext<EditReceiptActivity>) = with(ui) {
        verticalLayout {
            toolbar = toolbarTemplate {}
            scrollView {
                verticalLayout {
                    tableViewFrame {
                        textInputLayout {
                            editName = appCompatEditText {
                                id = R.id.act_edit_receipt_name
                                hint = resources.getString(R.string.name)
                            }
                        }
                        textInputLayout {
                            editDate = clickableView {
                                id = R.id.act_edit_receipt_date
                                hint = resources.getString(R.string.date)
                                onClick { ui.owner.onDate() }
                            }
                        }
                        textInputLayout {
                            editIncome = appCompatEditText {
                                id = R.id.act_edit_receipt_income
                                hint = resources.getString(R.string.income)
                                inputType = InputType.TYPE_CLASS_NUMBER
                            }
                        }
                        textInputLayout {
                            editSource = clickableView {
                                id = R.id.act_edit_receipt_source
                                hint = resources.getString(R.string.source)
                                onClick { ui.owner.onSource() }
                            }
                        }
                        textInputLayout {
                            editAccount = clickableView {
                                id = R.id.act_edit_receipt_account
                                hint = resources.getString(R.string.account)
                                onClick { ui.owner.onAccount() }
                            }
                        }
                        checkShowInResume = checkBox {
                            id = R.id.act_edit_receipt_show_in_resume
                            hint = resources.getString(R.string.show_in_resume)
                            isChecked = true
                        }.lparams {
                            bottomMargin = dip(5)
                        }
                        textInputLayout {
                            editRepetition = appCompatEditText {
                                id = R.id.act_edit_receipt_repetition
                                hint = resources.getString(R.string.repetition)
                                inputType = InputType.TYPE_CLASS_NUMBER
                                setText(R.string._1)
                            }
                        }
                        textInputLayout {
                            editInstallment = appCompatEditText {
                                id = R.id.act_edit_receipt_installment
                                hint = resources.getString(R.string.installments)
                                inputType = InputType.TYPE_CLASS_NUMBER
                                setText(R.string._1)
                            }
                        }
                    }
                }
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}
